eknrenderer/eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1312 @@
+                                                           source_name,
+                                                           error);
+      g_autofree char *license_link = NULL;
+      g_autofree char *disclaimer = NULL;
+
+      if (original_link == NULL)
+        return NULL;
+
+      license_link = format_license_link (license, error);
+
+      if (license_link == NULL)
+        return NULL;
+
+      disclaimer = g_strdup_printf (_("This page contains content from %s, available under a %s license."),
+                                    original_link, license_link);
+      return g_variant_new_take_string (g_steal_pointer (&disclaimer));
+    }
+  else if (g_strcmp0 (source, "wikihow") == 0)
+    {
+      g_autofree char *wikihow_article_link = format_a_href_link (original_uri,
+                                                                  title,
+                                                                  error);
+      g_autofree char *wikihow_link = NULL;
+      g_autofree char *disclaimer = NULL;
+
+      if (wikihow_article_link == NULL)
+        return NULL;
+
+      wikihow_link = format_a_href_link (_("http://wikihow.com"),
+                                         wikihow_article_link,
+                                         error);
+
+      if (wikihow_link == NULL)
+        return NULL;
+
+      disclaimer = g_strdup_printf (_("See %s for more details, videos, pictures and attribution. Courtesy of %s, where anyone can easily learn how to do anything."),
+                                   wikihow_article_link,
+                                   wikihow_link);
+
+      return g_variant_new_take_string (g_steal_pointer (&disclaimer));
+    }
+
+  return g_variant_new_boolean (FALSE);
+}
+
+static GVariant *
+get_legacy_css_files (const char *source)
+{
+  const char * const empty_css_files[] = { NULL };
+
+  if (g_strcmp0 (source, "wikisource") == 0 ||
+      g_strcmp0 (source, "wikibooks") == 0 ||
+      g_strcmp0 (source, "wikipedia") == 0)
+    {
+      const char * const css_files[] = {
+        "wikimedia.css",
+        NULL
+      };
+      return g_variant_new_strv (css_files, -1);
+    }
+  else if (g_strcmp0 (source, "wikihow") == 0)
+    {
+      const char * const css_files[] = {
+        "wikihow.css",
+        NULL
+      };
+      return g_variant_new_strv (css_files, -1);
+    }
+
+  return g_variant_new_strv (empty_css_files, -1);
+}
+
+static GVariant *
+get_legacy_javascript_files (gboolean use_scroll_manager)
+{
+  g_autoptr(GPtrArray) javascript_files = g_ptr_array_new ();
+  const char *content_fixes_js = "content-fixes.js";
+  const char *hide_broken_images_js = "hide-broken-images.js";
+  const char *scroll_manager_js = "scroll-manager.js";
+
+  g_ptr_array_add (javascript_files, (gpointer) content_fixes_js);
+  g_ptr_array_add (javascript_files, (gpointer) hide_broken_images_js);
+
+  if (use_scroll_manager)
+    g_ptr_array_add (javascript_files, (gpointer) scroll_manager_js);
```

Comment:
I doubt you have to cast to `gpointer` explicitly all the strings in this function; you most likely can simply pass the strings directly to `g_ptr_array_add()`

Reply:
I think the issue here was that I'd get a warning about an implicit cast away of `const`. But I'll check without it.

Reply:
Yeah, I get a warning if I do that:

```
usr/include/glib-2.0/glib/garray.h:168:12: note: expected ‘gpointer {aka void *}’ but argument is of type ‘const char *’
 void       g_ptr_array_add                (GPtrArray        *array,
```

Reply:
`g_ptr_array_add (javascript_files, (gpointer) "the-file.js");` should be the warning-less way to do it I'm looking for.

Reply:
Hum, that does work - TIL.